Controller/Notification/Index.php

```diff
@@ -76,42 +61,46 @@ public function execute()
     {
         $jsonContent = $this->request->getContent();
 
-        if (!$this->isValidSignature($jsonContent)) {
+        if (!$this->helper->isValidSignature($this->getSignature(), $jsonContent)) {
             return $this->setResponse(400, __('Failed comparison of CR-Signature and shop hash.'));
         }
 
-        $content = $this->serializer->unserialize($jsonContent);
+        $data = $this->serializer->unserialize($jsonContent);
 
-        if (!isset($content['externalId'])) {
-            return $this->setResponse(400, __('Empty content.'));
+        if (!isset($data['externalId'])) {
+            return $this->setResponse(400, 'External ID must be set.');
         }
 
-        $this->statusManagement->changeApplicationAndOrderStatus($content['externalId'], $content['status']);
+        if (!isset($data['status'])) {
+            return $this->setResponse(400, 'Status must be set.');
+        }
+
+        if ($data['status'] === ComfinoStatusManagement::CANCELLED_BY_SHOP) {
+            return $this->setResponse(400, 'Invalid status ' . ComfinoStatusManagement::CANCELLED_BY_SHOP . '.');
+        }
+
+        if (!$this->statusManagement->changeApplicationAndOrderStatus($data['externalId'], $data['status'])) {
+            return $this->setResponse(400, sprintf('Invalid status %s.', $data['status']));
+        }
 
         return $this->setResponse(200, 'OK');
     }
 
-    /**
-     * @param int $code
-     * @param string $content
-     *
-     * @return Json
-     */
-    private function setResponse(int $code, string $content): Json
+    private function getSignature(): string
     {
-        return $this->resultJsonFactory->create()->setHttpResponseCode($code)->setData(['status' => $content]);
+        $crSignature = $this->request->getHeader('CR-Signature');
+
+        if (!empty($crSignature)) {
+            return $crSignature;
+        }
+
+        $crSignature = $this->request->getHeader('X-CR-Signature');
```

Comment:
na 2 sposoby wysyłamy tego headera?

Reply:
Tak. To pomysł Pawła - niektóre serwery podobno odrzucają niestandardowe nagłówki i ignorują żądania z nimi, a przyjmują tylko z prefiksem X-, więc i po stronie wtyczki muszę uwzględnić taki przypadek. Notyfikacje najpierw wysyłają bez X-, a jak jest określony błąd, to ponawiana jest notyfikacja z nagłówkami z prefiksem X-.